File: probe/collectors/memory.py
import logging
import psutil

logger = logging.getLogger("pcmonitor.collectors.memory")


def collect():
    try:
        vm = psutil.virtual_memory()
        swap = psutil.swap_memory()
        return {
            "ram_percent": vm.percent,
            "ram_used_gb": round(vm.used / (1024 ** 3), 2),
            "ram_total_gb": round(vm.total / (1024 ** 3), 2),
            "swap_percent": swap.percent,
            "swap_used_gb": round(swap.used / (1024 ** 3), 2),
            "swap_total_gb": round(swap.total / (1024 ** 3), 2),
        }
    except Exception as e:
        logger.error(f"Memory collection error: {e}")
        return {
            "ram_percent": None, "ram_used_gb": None, "ram_total_gb": None,
            "swap_percent": None, "swap_used_gb": None, "swap_total_gb": None,
        }

File: probe-rust/src/collectors/memory.rs
use sysinfo::System;

pub struct MemInfo {
    pub ram_percent: f32,
    pub ram_used_gb: f64,
    pub ram_total_gb: f64,
    pub swap_percent: f32,
    pub swap_used_gb: f64,
}

pub fn collect(sys: &System) -> MemInfo {
    let total = sys.total_memory();
    let used = sys.used_memory();
    let ram_percent = if total > 0 {
        (used as f64 / total as f64 * 100.0) as f32
    } else {
        0.0
    };

    let swap_total = sys.total_swap();
    let swap_used = sys.used_swap();
    let swap_percent = if swap_total > 0 {
        (swap_used as f64 / swap_total as f64 * 100.0) as f32
    } else {
        0.0
    };

    const GB: f64 = 1_073_741_824.0;

    MemInfo {
        ram_percent,
        ram_used_gb: used as f64 / GB,
        ram_total_gb: total as f64 / GB,
        swap_percent,
        swap_used_gb: swap_used as f64 / GB,
    }
}
